solidity/random-beacon/contracts/libraries/Groups.sol

```diff
@@ -148,29 +127,194 @@ library Groups {
+
+        uint64 selectedGroup = uint64(seed % numberOfActiveGroups(self));
+        uint64 result = shiftByTerminatedGroups(
+            self,
+            shiftByExpiredGroups(self, selectedGroup)
+        );
+        return result;
     }
 
-    function selectGroup(Data storage self, uint256 seed)
+    /// @notice Setter for relay entry timeout.
+    /// @param timeout Relay entry timout calculated in Relay.sol lib.
+    function setRelayEntryTimeout(Data storage self, uint256 timeout) internal {
+        self.relayEntryTimeout = timeout;
+    }
+
+    /// @notice Setter for group lifetime.
+    /// @param lifetime Lifetime of a group in blocks.
+    function setGroupLifetime(Data storage self, uint256 lifetime) internal {
+        self.groupLifetime = lifetime;
+    }
+
+    /// @notice Checks if group with the given index is terminated.
+    function isGroupTerminated(Data storage self, uint64 groupId)
         internal
         view
-        returns (uint64)
+        returns (bool)
+    {
+        return self.groupsData[self.groupsRegistry[groupId]].terminated;
+    }
+
+    /// @notice Gets the cutoff time until which the given group is considered
+    ///         to be active assuming it hasn't been terminated before.
+    function groupLifetimeOf(Data storage self, bytes32 groupPubKeyHash)
+        internal
+        view
+        returns (uint256)
+    {
+        return
+            self.groupsData[groupPubKeyHash].activationBlockNumber +
+            self.groupLifetime;
+    }
+
+    /// @notice Gets the cutoff time in blocks after which the given group is
+    ///         considered as stale. Stale group is an expired group which is no
+    ///         longer performing any operations.
+    function groupStaleTime(Data storage self, bytes32 groupPubKeyHash)
+        internal
+        view
+        returns (uint256)
+    {
+        return groupLifetimeOf(self, groupPubKeyHash) + self.relayEntryTimeout;
+    }
+
+    /// @notice Checks if a group with the given index is a stale group.
+    ///         Stale group is an expired group which is no longer performing any
+    ///         operations. It is important to understand that an expired group
+    ///         may still perform some operations for which it was selected when
+    ///         it was still active. We consider a group to be stale when it's
+    ///         expired and when its expiration time and potentially executed
+    ///         operation timeout are both in the past.
+    function isStaleGroup(Data storage self, bytes memory groupPubKey)
+        internal
+        view
+        returns (bool)
+    {
+        // TODO: Can a group be considered as "stale" if "expiredGroupOffset" was
+        //       not moved forward which means that group selection wasn't triggered?
+        //       In other words can it be stale when it's not officially expired?
+        return groupStaleTime(self, keccak256(groupPubKey)) < block.number;
```

Comment:
I suggest we completely remove the concept of a stale group. Then, we do not need to pass `relayEntryTimeout` to this library. For `reportUnauthorizedSigning` let's assume no key can ever leak. 

This will simplify everything and no key ever leaked is a fair expectation.

Reply:
Ok, it will certainly simplify things. The stale functionality was used along with rewards in v1. If the rewards structure takes this into account, then I think we are fine. 921f30b